aea/runtime.py

```diff
@@ -0,0 +1,238 @@
+        )  # TODO: decide who constructs loop: agent, runtime, multiplexer.
+        self._state: AsyncState = AsyncState(RuntimeStates.initial)
+
+    def start(self) -> None:
+        """Start agent using runtime."""
+        if self._state.get() is RuntimeStates.started:
+            logger.error("Already started!")
+            return
+        self._start()
+
+    def stop(self) -> None:
+        """Stop agent and runtime."""
+        logger.debug("Runtime stop called!")
+        self.teardown()
+        self._stop()
+        logger.debug("[{}]: Stopped".format(self._agent.name))
+
+    def teardown(self) -> None:
+        """Tear down agent."""
+        logger.debug("Runtime teardown ...")
+        self._agent.teardown()
+        logger.debug("[{}]: Teardown completed".format(self._agent.name))
+
+    @abstractmethod
+    def _start(self) -> None:
+        """Implement runtime start function here."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def _stop(self) -> None:
+        """Implement runtime stop function here."""
+        raise NotImplementedError
+
+
+class AsyncRuntime(BaseRuntime):
+    """Asynchronous runtime: uses asyncio loop for multiplexer and async agent main loop."""
+
+    def __init__(
+        self, agent: "Agent", loop: Optional[AbstractEventLoop] = None
+    ) -> None:
+        """
+        Init runtime.
+
+        :param agent: Agent to run.
+        :param loop: optional event loop. if not provided a new one will be created.
+        :return: None
+        """
+        super().__init__(agent=agent, loop=loop)
+        self._stopping_task: Optional[asyncio.Task] = None
+        self._async_stop_lock: Optional[asyncio.Lock] = None
+
+    def _start(self) -> None:
+        """Start runtime."""
+        try:
+            if self._state.get() is RuntimeStates.started:
+                raise ValueError("Runtime alrady started!")
+
+            asyncio.set_event_loop(self._loop)
+            self._agent._multiplexer.set_loop(self._loop)
+            self._agent._main_loop.set_loop(self._loop)
+
+            self._state.set(RuntimeStates.started)
+            self._thread = threading.current_thread()
+            self._async_stop_lock = asyncio.Lock()
+
+            logger.debug(f"Start runtime event loop {self._loop}: {id(self._loop)}")
+            self._task = self._loop.create_task(self._start_coro())
+            self._loop.run_until_complete(self._task)
+            logger.debug("runtime loop stopped!")
+            self._stopping_task = None
+        except Exception:
+            logger.exception("During runtime processing")
+            raise
+
+    async def _start_coro(self) -> None:
+        """Implmement main loop of runtime."""
+        try:
+            await self._start_multiplexer()
+            self._agent._start_setup()
+            await self._agent._main_loop._start_coro()
+        except Exception:
+            logger.exception("AsyncRuntime exception during run:")
+            raise
```

Comment:
should we re-raise a specific exception here, like `AsyncRuntimeException`?

Reply:
not think so, I expect `AEAException` here, or general exception if something goes really wrong. there is no specific exception related to runtime should be raised during runtime execution.